feat(components): [config-provider] add global icons configuration

feynmx opened the pull request:
- Global icons added to dialog, dropdown, message, notification, select, and tour components.

#20755 

Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

feynmx committed:
feat(components): [config-provider] add global icons configuration

- Global icons added to dialog, dropdown, message, notification, select, and tour components.

btea commented on docs/en-US/component/config-provider.md:
```diff
@@ -41,6 +41,14 @@ config-provider/message
 
 :::
 
+## Icons Configurations
```
```suggestion
## Icons Configurations ^(2.9.11)
```

btea reviewed (commented):
Overall, there is no problem, but the newly added attributes need to add corresponding version tags.

feynmx committed:
docs: update docs/en-US/component/config-provider.md

Co-authored-by: btea <2356281422@qq.com>

feynmx committed:
feat(components): [config-provider] add clear icons configuration

- Updated select and select-v2 components to utilize a configurable clear icon.
- Added new time-picker and time-select components with clear icon support.
- Improved example documentation for config-provider to showcase new icons and components.
- Refactored related code to ensure consistent icon usage across components.

feynmx committed:
feat(components): [config-provider] enhance global icon configuration

- Updated icons.vue example to improve layout and showcase new icon configurations.
- Refactored alert, message, message-box, and notification components to utilize global icons.
- Added new icons for success, info, warning, and error to the global icons configuration.
- Improved type handling for icons in message and notification components.

feynmx committed:
refactor(components): [time-picker] update clear icon reference

feynmx committed:
docs: update icons configuration version to 2.9.12 in config-provider

Dsaquel commented on docs/en-US/component/config-provider.md:
```diff
@@ -116,6 +125,13 @@ In this section, you can learn how to use Config Provider to provide experimenta
+
+| Attribute | Description                                               | Type                     | Default |
+| --------- | --------------------------------------------------------- | ------------------------ | ------- |
+| close     | Custom close icon component for dialog, message, etc.     | ^[string] / ^[Component] | —       |
```
These attributes should should have a bump version as @btea said.
Also the `dropdown` property is a little bit confusing. Maybe something like `selectSuffix` or `{ select: { suffix: MyIcon }}` could fit well.

cn-xufei replied:
Thanks for the suggestion!

I get that dropdown might not be super intuitive at first, but I wanted to avoid suffix since it could be confused with the existing slot or prop. Also, the icon is used in dropdown-like behaviors across multiple components, so dropdown feels like a more general and reusable name.

To help clarify things, I added a mapping table and also updated related components’ descriptions to mention Config Provider `icons`

Dsaquel reviewed (commented):
I think it's a good idea to have a global icon provider :+1: .
Maybe we should list all components that need to inherit from it.

feynmx committed:
docs: update component documentation with consistent icon attributes

- Update icon attributes in Config Provider, Dialog, Message Box, Notification, Select, and Tour
- Add icon reference table in Config Provider example
- Standardize description format across components

Dsaquel commented on packages/components/config-provider/src/hooks/use-global-config.ts:
```diff
@@ -72,6 +89,21 @@ export function useGlobalComponentSettings(
   }
 }
 
+const defaultIcons = {
+  close: markRaw(Close),
+  dropdown: markRaw(ArrowDown),
+  clear: markRaw(CircleClose),
+  success: markRaw(SuccessFilled),
+  info: markRaw(InfoFilled),
+  warning: markRaw(WarningFilled),
+  error: markRaw(CircleCloseFilled),
+}
+
+export function useGlobalIcons() {
+  const globalConfig = useGlobalConfig('icons', defaultIcons)
+  return computed(() => Object.assign(defaultIcons, globalConfig.value))
+}
```
If there is no `globalConfig` we will don't have these default icons.
We should keep handling the default value as their place :+1: .

Dsaquel commented on docs/en-US/component/dialog.md:
```diff
@@ -145,36 +145,36 @@ dialog/events
+| center                     | whether to align the header and footer in center                                                                                                 | ^[boolean]                          | false   |
+| align-center ^(2.2.16)     | whether to align the dialog both horizontally and vertically                                                                                     | ^[boolean]                          | false   |
+| destroy-on-close           | destroy elements in Dialog when closed                                                                                                           | ^[boolean]                          | false   |
+| close-icon                 | custom close icon component, default value comes from [Config Provider icons](/en-US/component/config-provider#icons-attribute) `close` property | ^[string] / ^[Component]            | Close   |
```
Default values should be handled by their components.
No need these changes in the components doc, it should be clear from his context.

cn-xufei replied:
Thanks for your feedback.

My initial thought was that since component `props` have higher priority, it's better to let `ConfigProvider` handle the default values, so that we can keep the components themselves more lightweight and focused on override behavior rather than managing defaults.😊

Dsaquel replied:
I understand that, but in order to respect other properties handled by config provider button, link, etc... default values are handled by the end component. Button, link, etc... have no relation with config provider until we create it.

IMO, If you want to handle default values in directly inside config provider then it should be the same for all other component properties handled by config provider to keep consistency. And if so you can open additional PR to suggest this change.

cn-xufei replied:
Thanks for your input!

I checked the Link component — it also sets default values to undefined, and follows the same priority: component > global (ConfigProvider). So this approach is consistent with existing components.

Let me know if you see any specific inconsistency or potential improvement — happy to discuss!

Dsaquel replied:
>I checked the Link component — it also sets default values to undefined, and follows the same priority: component > global (ConfigProvider). So this approach is consistent with existing components.
>
With the current changes, the default value is controlled in `useGlobalIcons`, which is different from the way Link (and others) components: they let the component control the default value, not the hook.

I understand the idea of putting defaults in `useGlobalIcons` and making it work like hooks such as `useLocale` or `useEmptyValues`. However, those hooks have more complex logic, whereas `globalIcons` is only about forwarding values.

For these reasons, and because there’s no additional logic in the hook, I believe we should follow the same pattern as Link, Button, etc. — letting the component define the default value, and using the hook only to inject the global config if present.

What do you think?

Dsaquel commented on packages/components/message-box/src/index.vue:
```diff
@@ -314,15 +320,19 @@ export default defineComponent({
-      return state.icon || (type && TypeComponentsMap[type]) || ''
+      return (
+        state.icon ||
+        (type && globalIcons.value[type === 'primary' ? 'info' : type]) ||
```
`TypeComponentsMap` is seems useful if there is no globalConfig.

cn-xufei replied:
Thanks for the feedback, I replaced TypeComponentsMap with globalIcons to leverage ConfigProvider for consistent icon defaults, Even without ConfigProvider, globalIcons is always available by default, ensuring reliability.  😊

feynmx committed:
chore: remove unused eslint working directories configuration

feynmx committed:
refactor(components): [config-provider] add missing icon types

feynmx committed:
docs: update icons configuration version to 2.10.2 in config-provider

feynmx committed:
refactor(components): [config-provider] standardize icon handling

- Updated icon refs in Dialog, MessageBox, Notification, Select to use global config
- Improved icon type handling in Message and Notification
- Unified clear icon usage in Select and Time Picker
- Enhanced icon attribute docs for affected components

feynmx committed:
fix(components): update notification and select tests for icon handling

- Removed unused Close icon import in Notification tests and updated close button existence check.
- Added console log in Select tests for debugging purposes.
- Set default suffix icon to ArrowDown in Select component.

feynmx committed:
fix(components): update notification component to use Close icon

- Added Close icon import in Notification component and set it as the default close icon.
- Updated notification tests to check for the existence of the Close icon component.
- Adjusted logic in notification.vue to prioritize global close icon configuration.

feynmx committed:
fix(components): add default Close icon to Dialog and Message

- Imported Close icon in Dialog and Message components, setting it as the default close icon.
- Updated logic in Dialog and Message to prioritize global icon configuration for close icons.

feynmx committed:
refactor(components): update icon handling in Notification and Tour

- Replaced usage of global icon configuration with iconsConfig in Notification and Tour components.
- Updated logic to ensure the correct close icon is utilized from the new iconsConfig.

feynmx committed:
refactor(components): rename dropdown icon to arrowDown

- Updated icon references in config-provider to replace 'dropdown' with 'arrowDown'.
- Adjusted related components (Dropdown, Select) to utilize the new icon naming convention.
- Ensured consistency across icon configurations for better clarity and usage.

feynmx committed:
refactor(components): enhance close icon handling in Drawer

- Replaced static Close icon with a dynamic component for better flexibility.
- Integrated global icon configuration to allow for customizable close icons.
- Updated logic to compute the close icon component based on global settings or props.

Dsaquel commented on packages/components/config-provider/src/config-provider-props.ts:
```diff
@@ -12,6 +12,16 @@ export type ExperimentalFeatures = {
 
+export type IconConfigContext = {
+  close?: string | Component
+  arrowDown?: string | Component
```
thanks for changing `dropdown` to `arrowDown`, please can you sync the change in the doc.

Dsaquel commented on packages/components/config-provider/src/hooks/use-global-config.ts:
```diff
@@ -72,6 +73,10 @@ export function useGlobalComponentSettings(
 }
 
+export function useGlobalIcons() {
+  return useGlobalConfig('icons', {} as IconConfigContext)
```
I think we can use directly `useGlobalConfig('icons')` into the end components.
I don't think we need `{} as IconConfigContext` as well.

Dsaquel commented on packages/components/select/__tests__/select.test.ts:
```diff
@@ -1090,6 +1090,7 @@ describe('Select', () => {
     })
 
     const suffixIcon = wrapper.findComponent(ArrowDown)
+    console.log(wrapper.html())
```
```suggestion
```

Dsaquel reviewed (commented):
overall it is good.

I'll test a bit deeper later :+1:

feynmx committed:
refactor(components): enhance global icon handling across components

- Use optional chaining for global icons in Alert, Drawer, Dropdown, Select, and Time Picker
- Rename dropdown icon to `arrowDown` for consistency
- Add fallback to default props when global icons are missing

cn-xufei commented on packages/components/message-box/src/index.vue:
```diff
@@ -283,7 +290,7 @@ export default defineComponent({
       distinguishCancelAndClose: false,
       icon: '',
-      closeIcon: '',
+      closeIcon: props.closeIcon as string | Component,
```
It’s also fine not to include it in props; the original intention was to maintain consistency with the default props behavior across other components.

cn-xufei commented on packages/components/message-box/src/index.vue:
```diff
@@ -62,7 +62,7 @@
               >
                 <el-icon :class="ns.e('close')">
-                  <component :is="closeIcon || 'close'" />
+                  <component :is="closeIconComponent" />
```
> props have higher priority than iconsConfig.

Initially, props had higher priority, but to ensure the default values of the component props are preserved, the current logic gives higher precedence to iconsConfig over props, as I understand it.

feynmx committed:
refactor(components): replace useGlobalIcons with useGlobalConfig

- Remove useGlobalIcons function from use-global-config.ts
- Update all components to use useGlobalConfig('icons') instead
- Affects alert, dialog, drawer, dropdown, select, and time-picker components
